qiskit/optimization/problems/optimization_problem.py

```diff
@@ -342,227 +345,464 @@ def set_problem_type(self, _type):
+                correctly. The lower and upper bounds are updated accordingly.
+                i.e., SparseTriple.ind1 (variable)
+                        -> SparseTriple.ind2 (variable) * SparseTriple.val (constant)
+
+        Returns:
+            a tuple of `OptimizationProblem` and `SubstituteStatus`.
+            If the resulting problem has no issue, the status is `ok`.
+            Otherwise, an empty problem and status `infeasible` are returned.
 
-        constants: SparsePair (replace variable by constant)
-        variables: SparseTriple (replace variables by weighted other variable
-        need to copy everything using name reference to make sure that indices are matched correctly
+        Raises:
+            QiskitOptimizationError: if the substitution is invalid as follows.
+                - Same variable is substituted multiple times.
+                - Coefficient of variable substitution is zero.
+
+        Example usage:
+
+        >>> from qiskit.optimization import OptimizationProblem
+        >>> from cplex import SparsePair, SparseTriple
+        >>> op = OptimizationProblem()
+        >>> op.variables.add(names=['x', 'y'], types='I'*2, lb=[-1]*2, ub=[2]*2)
+        >>> op.objective.set_sense(op.objective.sense.minimize)
+        >>> op.objective.set_linear([('x', 1), ('y', 2)])
+        >>> op.linear_constraints.add(lin_expr=[(['x', 'y'], [1.0, -1.0])], senses=['L'], rhs=[1.0])
+        >>> print(op.write_as_string())
+        \\ENCODING=ISO-8859-1
+        \\Problem name:
+
+        Minimize
+         obj1: x + 2 y
+        Subject To
+         c1: x - y <= 1
+        Bounds
+        -1 <= x <= 2
+        -1 <= y <= 2
+        Generals
+         x  y
+        End
+        >>> # substitute x <- 2
+        >>> op2, st = op.substitute_variables(constants=SparsePair(ind=['x'], val=[2]))
+        >>> print(st)
+        SubstituteStatus.ok
+        >>> print(op2.write_as_string())
+        \\ENCODING=ISO-8859-1
+        \\Problem name:
+
+        Minimize
+         obj1: 2 y + 2
+        Subject To
+         c1: - y <= -1
+        Bounds
+        -1 <= y <= 2
+        Generals
+         y
+        End
+        >>> # substitute y <- -x
+        >>> op3, st = op.substitute_variables(variables=SparseTriple(\
+                                                ind1=['y'], ind2=['x'], val=[-1]))
+        >>> print(st)
+        SubstituteStatus.ok
+        >>> print(op3.write_as_string())
+        \\ENCODING=ISO-8859-1
+        \\Problem name:
+
+        Minimize
+         obj1: - x
+        Subject To
+         c1: 2 x <= 1
+        Bounds
+        -1 <= x <= 1
+        Generals
+         x
+        End
+        """
+        subs = SubstituteVariables()
+        return subs.substitute_variables(src=self, constants=constants, variables=variables)
+
+
+class SubstituteStatus(Enum):
+    ok = 1
+    infeasible = 2
+
+
+class SubstituteVariables(object):
```

Comment:
Classes implicitly inherit from `object`, that can be skipped.
```suggestion
class SubstituteVariables:
```

Reply:
done